### PR TITLE
Update to default QS exemptions

### PR DESCRIPTION
**Pull request by jacobd91**
In a previous update a hard-coded list was added to the accepted QS arguments that couldn't be modified via a setting. This PR migrates that list from a hard-coded implementation to merge it into the accepted QS setting.

This includes setting said list as the default for new users and merging it with existing values for users that set values.

Additionally an "Add Defaults" button has been added next to the accepted QS textarea under the Page Cache advanced settings that will merge said list with existing values.

### CODE REVIEW

**Review by cssjoe — CHANGES_REQUESTED** [review]
See the review comments.

File: Generic_Plugin_Admin.php [hunk 1]
```diff
@@ -404,17 +404,23 @@ public function admin_head() {
...
 
-			if ( ! $config->get_boolean( 'pgcache.enabled' ) && $state_master->get_integer( 'common.install' ) > strtotime( 'NOW - 1 WEEK' ) ) {
+			if ( ! $this->_config->get_boolean( 'pgcache.enabled' ) && $state_master->get_integer( 'common.install' ) > strtotime( 'NOW - 1 WEEK' ) ) {
 				wp_safe_redirect( esc_url( network_admin_url( 'admin.php?page=w3tc_setup_guide' ) ) );
 			}
 		}
 
+		if ( empty( $this->_config->get_integer( 'pgcache.migrated.qsexempts' ) ) ) {
+			$pgcache_accept_qs = array_unique( array_merge( $this->_config->get_array( 'pgcache.accept.qs' ), PgCache_QsExempts::get_qs_exempts() ) );
+			$this->_config->set( 'pgcache.accept.qs', ksort( $pgcache_accept_qs ) );
```
[the remark this answers is not in the document]
Comment (jacobd91) [reply Generic_Plugin_Admin.php:419]: Updated

File: PgCache_Page_View.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+ * @since 2.1.0
+ *
+ * @global W3TCPgCacheQsExemptsData
+ */
+
+jQuery(function() {
+	jQuery('.pgcache_qsexempts_default').on(
+		'click',
+		function(){
+			var $pgcacheQsExempts = jQuery('#pgcache_accept_qs').val().split("\n");
```
[the remark this answers is not in the document]
Comment (jacobd91) [reply PgCache_Page_View.js]: Removed the $

File: inc/options/pgcache.php [hunk 3]
```diff
@@ -610,6 +610,7 @@
 					<textarea id="pgcache_accept_qs" name="pgcache__accept__qs"
 						<?php Util_Ui::sealing_disabled( 'pgcache.' ); ?>
 						cols="40" rows="5"><?php echo esc_textarea( implode( "\r\n", $this->_config->get_array( 'pgcache.accept.qs' ) ) ); ?></textarea>
+					<input type="button" class="button pgcache_qsexempts_default" value="Add Defaults" />
```
[the remark this answers is not in the document]
Comment (jacobd91) [reply inc/options/pgcache.php]: Updated to prefix with w3tc- and to replace _ with -

The rest of the file classes are not this format however.

File: Generic_AdminActions_Default.php [hunk 4]
```diff
@@ -273,6 +273,36 @@ private function _w3tc_save_options_process() {
...
+				'pgcache.reject.authors',
+				'pgcache.reject.custom',
+				'pgcache.accept.files',
+				'pgcache.accept.uri',
+				'pgcache.cache.headers',
+			);
+			// Remove empty lines, trim values, and sort.
+			foreach ( $settings as $setting ) {
+				if ( ! empty( $config->get_array( $setting ) ) ) {
+					$values = array_map( 'trim', array_filter( $config->get_array( $setting ) ) );
```
Comment (cssjoe) [inline Generic_AdminActions_Default.php:299]: I don't see a callback function in array_filter().  Maybe you were trying to use it like this example:
https://www.php.net/manual/en/function.array-filter.php#114931
```
$value = $config->get_array( $setting );

if ( ! empty( $value ) ) {
    $array = array_filter( array_map( 'trim', $value ), 'strlen' );
```
That will trim all array values and then remove empty lines.

**Review by cssjoe — CHANGES_REQUESTED** [review]

**Review by cssjoe — APPROVED** [review]

**Pull request merged**